In [85]:
import pandas as pd
from sklearn.metrics import precision_score
from sklearn.metrics import accuracy_score
from sklearn.metrics import recall_score
from sklearn.metrics import classification_report

1. Create a new file named model_evaluation.py or model_evaluation.ipynb for these exercises.

2.Given the following confusion matrix, evaluate (by hand) the model's performance.


|               | pred dog   | pred cat   |
|:------------  |-----------:|-----------:|
| actual dog    |         46 |         7  |
| actual cat    |         13 |         34 |


In the context of this problem, what is a false positive?

In [21]:
#positive case = dog
#negative case = cat

#if dog is our positive case, false positive is when your predict a dog but it was actually a cat
#which is 13 on this case

In the context of this problem, what is a false negative

In [22]:
#if dog is our positive case, false negative is prediction a cat but it was a dog actually which is 7

In [23]:
tp=46
tn=34
fp=13
fn=7

How would you describe this model?

In [24]:
precision = tp/(tp+fp)
accuracy = (tp+tn)/(tp+tn+tp+fp)
recall = tp/(tp+fn)
print("positive case = dog")
print('precision', round(precision,2))
print('accuracy', round(accuracy,2))
print('recall', round(recall,2))

positive case = dog
precision 0.78
accuracy 0.58
recall 0.87


3.You are working as a datascientist working for Codeup Cody Creator (C3 for short), a rubber-duck manufacturing plant.

Unfortunately, some of the rubber ducks that are produced will have defects. Your team has built several models that try to predict those defects, and the data from their predictions can be found here.

Use the predictions dataset and pandas to help answer the following questions:

An internal team wants to investigate the cause of the manufacturing defects. They tell you that they want to identify as many of the ducks that have a defect as possible. Which evaluation metric would be appropriate here? Which model would be the best fit for this use case?

In [ ]:
#postive case = defective duck

In [89]:
#load csv
df = pd.read_csv("c3.csv")
df.head()

actual     model1  model2     model3
0  No Defect  No Defect  Defect  No Defect
1  No Defect  No Defect  Defect     Defect
2  No Defect  No Defect  Defect  No Defect
3  No Defect     Defect  Defect     Defect
4  No Defect  No Defect  Defect  No Defect

In [90]:
# not only true postives(defective ducks), we need to track down ducks we think are good but are infact defective,
#which is false negative. so we use recall here. 


In [99]:
#we can use classification report on all there models
print('model1')
print(pd.DataFrame(met.classification_report(df.actual, df.model1, output_dict=True)))
print()
print('model2')
print(pd.DataFrame(met.classification_report(df.actual, df.model2, output_dict=True)))
print()
print('model3')
print(pd.DataFrame(met.classification_report(df.actual, df.model3, output_dict=True)))

model1
              precision    recall  f1-score   support

      Defect       0.80      0.50      0.62        16
   No Defect       0.96      0.99      0.97       184

    accuracy                           0.95       200
   macro avg       0.88      0.74      0.79       200
weighted avg       0.95      0.95      0.94       200


model2
              precision    recall  f1-score   support

      Defect       0.10      0.56      0.17        16
   No Defect       0.94      0.56      0.70       184

    accuracy                           0.56       200
   macro avg       0.52      0.56      0.44       200
weighted avg       0.87      0.56      0.66       200


model3
              precision    recall  f1-score   support

      Defect       0.13      0.81      0.23        16
   No Defect       0.97      0.53      0.69       184

    accuracy                           0.56       200
   macro avg       0.55      0.67      0.46       200
weighted avg       0.90      0.56      0.65       2

In [ ]:
#since recall for model 3 is the highest,81.25% , we will use model 3

Recently several stories in the local news have come out highlighting customers who received a rubber duck with a defect, and portraying C3 in a bad light. The PR team has decided to launch a program that gives customers with a defective duck a vacation to Hawaii. They need you to predict which ducks will have defects, but tell you the really don't want to accidentally give out a vacation package when the duck really doesn't have a defect. Which evaluation metric would be appropriate here? Which model would be the best fit for this use case?

In [36]:
#here, postive case is defective ducks.
#pr team also wants to catch false postive, ie. giving out good ducks assuming it is defective
#so we use precision 

In [ ]:
as in the table, model1 has highest precision, 80% so we choose it

4. You are working as a data scientist for Gives You Paws ™, a subscription based service that shows you cute pictures of dogs or cats (or both for an additional fee).

At Gives You Paws, anyone can upload pictures of their cats or dogs. The photos are then put through a two step process. First an automated algorithm tags pictures as either a cat or a dog (Phase I). Next, the photos that have been initially identified are put through another round of review, possibly with some human oversight, before being presented to the users (Phase II).

Several models have already been developed with the data, and you can find their results here.

Given this dataset, use pandas to create a baseline model (i.e. a model that just predicts the most common class) and answer the following questions:

In [43]:
df_paws = pd.read_csv('gives_you_paws.csv')
df_paws.head()

actual model1 model2 model3 model4
0    cat    cat    dog    cat    dog
1    dog    dog    cat    cat    dog
2    dog    cat    cat    cat    dog
3    dog    dog    dog    cat    dog
4    cat    cat    cat    dog    dog

In [44]:
#to find baseline, do value counts of actual as baseline is the mode
df_paws.actual.value_counts()

dog    3254
cat    1746
Name: actual, dtype: int64

In [45]:
#so baseline is dog,make a column for it
df_paws["baseline"] = "dog"
df_paws.head()

actual model1 model2 model3 model4 baseline
0    cat    cat    dog    cat    dog      dog
1    dog    dog    cat    cat    dog      dog
2    dog    cat    cat    cat    dog      dog
3    dog    dog    dog    cat    dog      dog
4    cat    cat    cat    dog    dog      dog

a. In terms of accuracy, how do the various models compare to the baseline model? Are any of the models better than the baseline?

In [46]:
#we can use classification report but lets practice another way
#accuracy is looking at predicting true postive and true negative correction 
#between actual and model,so
model1_accuracy = (df_paws.actual == df_paws.model1).mean()
model2_accuracy = (df_paws.actual == df_paws.model2).mean()
model3_accuracy = (df_paws.actual == df_paws.model3).mean()
model3_accuracy = (df_paws.actual == df_paws.model3).mean()
model4_accuracy = (df_paws.actual == df_paws.model4).mean()
baseline_accuracy = (df_paws.actual == df_paws.baseline).mean()
model1_accuracy, model2_accuracy, model3_accuracy,model4_accuracy, baseline_accuracy

(0.8074, 0.6304, 0.5096, 0.7426, 0.6508)

as baseline accuray is 0.6508, only model 1 and model 4 has accuracy over
that value so model1, model 4 are better than baseline for accuracy

b. Suppose you are working on a team that solely deals with dog pictures. Which of these models would you recomend for Phase I? For Phase II?

In [48]:
# positive case = dog
#we need to minimize case where we say it is dog but actually is a cat
#ie. false postive. so we need to maximize precision
#make subset of dog in all model to comapre with actual
model1_subset = df_paws[df_paws.model1 == 'dog'] 
model2_subset = df_paws[df_paws.model2 == 'dog'] 
model3_subset = df_paws[df_paws.model3 == 'dog'] 
model4_subset = df_paws[df_paws.model4 == 'dog'] 
baseline_subset = df_paws[df_paws.baseline == "dog"]

In [51]:
#get precision
model1_precision = (model1_subset.actual == model1_subset.model1).mean()
model2_precision = (model2_subset.actual == model2_subset.model2).mean()
model3_precision = (model3_subset.actual == model3_subset.model3).mean()
model4_precision = (model4_subset.actual == model4_subset.model4).mean()
baseline_precision = (baseline_subset.actual == baseline_subset.baseline).mean()
round(model1_precision,4), round(model2_precision,4),round(model3_precision,4), round(model4_precision,4), round(baseline_precision,4)

(0.89, 0.8932, 0.6599, 0.7312, 0.6508)

In [ ]:
# we use precision to minimize false positive
# we can recommend model 1 and model 2 being alot higher than baseline

c. Suppose you are working on a team that solely deals with cat pictures. Which of these models would you recomend for Phase I? For Phase II?

In [52]:
# positive class = cat
#negative class = dog
#we need to minimize case where we say it is cat but actually is a dog
#ie. false postive. so we need to maximize precision
#make subset of cat in all model to comapre with actual
model1_subset = df_paws[df_paws.model1 == 'cat'] 
model2_subset = df_paws[df_paws.model2 == 'cat'] 
model3_subset = df_paws[df_paws.model3 == 'cat'] 
model4_subset = df_paws[df_paws.model4 == 'cat'] 
baseline_subset = df_paws[df_paws.baseline == "cat"]

In [53]:
#get precision
model1_precision = (model1_subset.actual == model1_subset.model1).mean()
model2_precision = (model2_subset.actual == model2_subset.model2).mean()
model3_precision = (model3_subset.actual == model3_subset.model3).mean()
model4_precision = (model4_subset.actual == model4_subset.model4).mean()
baseline_precision = (baseline_subset.actual == baseline_subset.baseline).mean()
round(model1_precision,4), round(model2_precision,4),round(model3_precision,4), round(model4_precision,4), round(baseline_precision,4)

(0.6898, 0.4841, 0.3583, 0.8072, nan)

In [ ]:
# we use precision to minimize false positive
#we chose model 4 being the largest value of all other model

5. Follow the links below to read the documentation about each function, then apply those functions to the data from the previous problem.

sklearn.metrics.accuracy_score

sklearn.metrics.precision_score

sklearn.metrics.recall_score

sklearn.metrics.classification_report



Use accuracy_score

In [55]:
# Model 1 Accuracy Score
y_true = df_paws.actual
y_pred = df_paws.model1
accuracy_score(y_true, y_pred)

0.8074

In [57]:
# Model 2 Accuracy Score
y_true = df_paws.actual
y_pred = df_paws.model2
accuracy_score(y_true, y_pred)

0.6304

In [58]:
# Model 3 Accuracy Score
y_true = df_paws.actual
y_pred = df_paws.model3
accuracy_score(y_true, y_pred)

0.5096

In [59]:
# Model 4 Accuracy Score
y_true = df_paws.actual
y_pred = df_paws.model4
accuracy_score(y_true, y_pred)

0.7426

Use Precision_Score

In [60]:
#create subset for precision for all model
#dog is postive case
model1_subset = df_paws[df_paws.model1 == 'dog'] 
model2_subset = df_paws[df_paws.model2 == 'dog'] 
model3_subset = df_paws[df_paws.model3 == 'dog'] 
model4_subset = df_paws[df_paws.model4 == 'dog'] 

In [63]:
# Model 1 Precision Score
y_true = model1_subset.actual
y_pred = model1_subset.model1
round(recall_score(y_true, y_pred, average='micro'),4)


0.89

In [66]:
# Model 2 Precision Score
y_true = model2_subset.actual
y_pred = model2_subset.model2
round(recall_score(y_true, y_pred, average='micro'),4)


0.8932

In [67]:
# Model 3 Precision Score
y_true = model3_subset.actual
y_pred = model3_subset.model3
round(recall_score(y_true, y_pred, average='micro'),4)

0.6599

In [68]:
# Model 4 Precision Score
y_true = model4_subset.actual
y_pred = model4_subset.model4
round(recall_score(y_true, y_pred, average='micro'),4)

0.7312

Use Recall_Score

In [76]:
df_paws.head()

actual model1 model2 model3 model4 baseline
0    cat    cat    dog    cat    dog      dog
1    dog    dog    cat    cat    dog      dog
2    dog    cat    cat    cat    dog      dog
3    dog    dog    dog    cat    dog      dog
4    cat    cat    cat    dog    dog      dog

In [77]:
#create subset for recall on actual 
#dog is positive case
subset_actual = df_paws[df_paws.actual == "dog"]
subset_actual.head()

actual model1 model2 model3 model4 baseline
1    dog    dog    cat    cat    dog      dog
2    dog    cat    cat    cat    dog      dog
3    dog    dog    dog    cat    dog      dog
5    dog    dog    dog    dog    dog      dog
8    dog    dog    cat    dog    dog      dog

In [78]:
# Model 1 Recall Score
y_true = subset_actual.actual
y_pred = subset_actual.model1
recall_score(y_true, y_pred, average='micro')

0.803318992009834

In [79]:
# Model 2 Recall Score
y_true = subset_actual.actual
y_pred = subset_actual.model2
recall_score(y_true, y_pred, average='micro')

0.49078057775046097

In [80]:
# Model 3 Recall Score
y_true = subset_actual.actual
y_pred = subset_actual.model3
recall_score(y_true, y_pred, average='micro')

0.5086047940995697

In [81]:
# Model 4 Recall Score
y_true = subset_actual.actual
y_pred = subset_actual.model4
recall_score(y_true, y_pred, average='micro')

0.9557467732022127

Use Classification_Report

In [95]:
# Model 1 Classification Report
y_true = df_paws.actual
y_pred = df_paws.model1
print(classification_report(y_true, y_pred))

              precision    recall  f1-score   support

         cat       0.69      0.82      0.75      1746
         dog       0.89      0.80      0.84      3254

    accuracy                           0.81      5000
   macro avg       0.79      0.81      0.80      5000
weighted avg       0.82      0.81      0.81      5000



In [100]:
# Model 2 Classification Report
y_true = df_paws.actual
y_pred = df_paws.model2
print(classification_report(y_true, y_pred))

              precision    recall  f1-score   support

         cat       0.48      0.89      0.63      1746
         dog       0.89      0.49      0.63      3254

    accuracy                           0.63      5000
   macro avg       0.69      0.69      0.63      5000
weighted avg       0.75      0.63      0.63      5000



In [101]:
# Model 3 Classification Report
y_true = df_paws.actual
y_pred = df_paws.model3
print(classification_report(y_true, y_pred))

              precision    recall  f1-score   support

         cat       0.36      0.51      0.42      1746
         dog       0.66      0.51      0.57      3254

    accuracy                           0.51      5000
   macro avg       0.51      0.51      0.50      5000
weighted avg       0.55      0.51      0.52      5000



In [102]:
# Model 4 Classification Report
y_true = df_paws.actual
y_pred = df_paws.model4
print(classification_report(y_true, y_pred))

              precision    recall  f1-score   support

         cat       0.81      0.35      0.48      1746
         dog       0.73      0.96      0.83      3254

    accuracy                           0.74      5000
   macro avg       0.77      0.65      0.66      5000
weighted avg       0.76      0.74      0.71      5000

